fe/src/main/java/org/apache/doris/planner/MaterializedViewSelector.java

```diff
@@ -0,0 +1,453 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.planner;
+
+import org.apache.doris.analysis.Analyzer;
+import org.apache.doris.analysis.CastExpr;
+import org.apache.doris.analysis.Expr;
+import org.apache.doris.analysis.FunctionCallExpr;
+import org.apache.doris.analysis.SelectStmt;
+import org.apache.doris.analysis.SlotRef;
+import org.apache.doris.analysis.TableRef;
+import org.apache.doris.catalog.Column;
+import org.apache.doris.catalog.KeysType;
+import org.apache.doris.catalog.OlapTable;
+import org.apache.doris.catalog.Table;
+import org.apache.doris.common.UserException;
+import org.apache.doris.qe.ConnectContext;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+
+/**
+ * The new materialized view selector supports SPJ<->SPJG.
+ * At the same time, it is compatible with all the features of the old version.
+ * The SPJ query is "Select Projection and Join" such as:
+ *     select t1.c1 from t1, t2 where t1.c2=t2.c2 and t1.c3=1;
+ * The SPJG query is "Select Projection Join and Group-by" such as:
+ *     select t1.c1, sum(t2.c1) from t1, t2 where t1.c2=t2.c2 and t1.c3=1 group by t1.c1;
+ */
+public class MaterializedViewSelector {
+    private static final Logger LOG = LogManager.getLogger(MaterializedViewSelector.class);
+
+    private final SelectStmt selectStmt;
+    private final Analyzer analyzer;
+
+    private Map<String, Set<String>> columnNamesInPredicates = Maps.newHashMap();
+    private boolean isSPJQuery;
+    private Map<String, Set<String>> columnNamesInGrouping = Maps.newHashMap();
+    private Map<String, Set<AggregatedColumn>> aggregateColumnsInQuery = Maps.newHashMap();
+    private Map<String, Set<String>> columnNamesInQueryOutput = Maps.newHashMap();
+
+    private boolean disableSPJGView;
+    private String reasonOfDisable;
+    private boolean isPreAggregation = true;
+
+    public MaterializedViewSelector(SelectStmt selectStmt, Analyzer analyzer) {
+        this.selectStmt = selectStmt;
+        this.analyzer = analyzer;
+        init();
+    }
+
+    /**
+     * There are two stages to choosing the best MV.
+     * Phase 1: Predicates
+     * According to aggregation and column information in the select stmt,
+     * the candidate MVs that meets the query conditions are selected.
+     * Phase 2: Priorities
+     * According to prefix index and row count in candidate MVs,
```

Comment:
I think method name `predicates` and `priorities` couldn't describe what we do, and we should change it.
Such as `computeCandidateMVs`, `computeBestMVByCost`.

Reply:
Actually, the `predicates` and `priorities` are carbonized from the kubernetes node selector.
These two names are more abstract.
Maybe I can add more comments to introduce it.

Reply:
OK